src/config.rs

```diff
@@ -111,55 +162,80 @@ pub struct DetectorConfig {
-        let mut config: OrchestratorConfig = serde_yml::from_str(&s)
-            .unwrap_or_else(|error| panic!("invalid orchestrator config: {error}"));
-        // Map tls name to tls config
-        // TODO: do this cleaner
+        let s = tokio::fs::read_to_string(path).await.map_err(Error::from)?;
+        serde_yml::from_str(&s).map_err(Error::from)
+    }
+
+    fn map_tls_configs(config: &mut Self) -> Result<(), Error> {
         let tls_configs = &config.tls;
-        config.generation.service =
-            service_tls_name_to_config(tls_configs, config.generation.service);
-        config.chunkers = config
-            .chunkers
-            .into_iter()
-            .map(|(name, mut config)| {
-                config.service = service_tls_name_to_config(tls_configs, config.service);
-                (name, config)
-            })
-            .collect::<HashMap<_, _>>();
-        config.detectors = config
-            .detectors
-            .into_iter()
-            .map(|(name, mut config)| {
-                config.service = service_tls_name_to_config(tls_configs, config.service);
-                (name, config)
-            })
-            .collect::<HashMap<_, _>>();
-        debug!(?config, "loaded orchestrator config");
-        config
+        match config.generation {
+            Some(ref mut generation) => {
+                generation.service =
+                    Self::tls_name_to_config(tls_configs, generation.clone().service)?;
+            }
+            None => warn!("no generation service configuration provided"),
+        }
+
+        match config.chunkers {
+            Some(ref mut chunkers) => {
+                for (_, chunker) in chunkers.iter_mut() {
+                    chunker.service =
+                        Self::tls_name_to_config(tls_configs, chunker.clone().service)?;
+                }
+            }
+            None => warn!("no chunker service configurations provided"),
+        }
+
+        for (_, detector) in config.detectors.iter_mut() {
+            detector.service = Self::tls_name_to_config(tls_configs, detector.clone().service)?;
+        }
+
+        Ok(())
     }
 
-    fn _validate(&self) {
-        todo!()
+    fn validate(&self) -> Result<Self, Error> {
+        if self.detectors.is_empty() {
+            Err(Error::NoDetectorsConfigured)
+        } else {
+            for (_, detector) in self.detectors.iter() {
+                if !self
+                    .clone()
+                    .chunkers
+                    .unwrap()
+                    .contains_key(&detector.chunker_id)
+                    && detector.chunker_id != DEFAULT_MODEL_ID
+                {
+                    return Err(Error::DetectorChunkerNotFound {
+                        detector: detector.clone().service.hostname,
+                        chunker: detector.chunker_id.clone(),
+                    });
+                }
+            }
+            Ok(self.clone())
+        }
     }
```

Comment:
Can we change this to return `Result<(), Error>`? It's probably more of a style preference, but the clone can also be avoid.

Suggested code below including a few tweaks.
1. Returns `Result<(), Error>` instead of cloning self
2. `self.clone().chunkers.unwrap()` has an unnecessary clone and would panic if no chunkers are configured
3. We should check for a default `chunker_id` _or_ a valid configured `chunker_id`
4. In the error, clone `detector.service.hostname` instead of `detector`

```suggestion
    fn validate(&self) -> Result<(), Error> {
        if self.detectors.is_empty() {
            Err(Error::NoDetectorsConfigured)
        } else {
            for detector in self.detectors.values() {
                // Chunker is valid
                let valid_chunker = detector.chunker_id == DEFAULT_MODEL_ID
                    || self
                        .chunkers
                        .as_ref()
                        .is_some_and(|chunkers| chunkers.contains_key(&detector.chunker_id));
                if !valid_chunker {
                    return Err(Error::DetectorChunkerNotFound {
                        detector: detector.service.hostname.clone(),
                        chunker: detector.chunker_id.clone(),
                    });
                }
            }
            Ok(())
        }
    }
```

Update `OrchestratorConfig::load()` accordingly:
```rust
//...
config.validate()?;
Ok(config)
```